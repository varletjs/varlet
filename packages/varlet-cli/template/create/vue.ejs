<template>
  <div class="var-<%- kebabCaseName %>"></div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'
import './<%- kebabCaseName %>.less'

export default defineComponent({
  name: '<%- namespace + bigCamelizeName %>',
})
</script>
