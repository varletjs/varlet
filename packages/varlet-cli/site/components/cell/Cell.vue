<template>
  <div :class="classes(n(), [border, n('--border')])">
    <div :class="classes(n('icon'), [iconClass, iconClass])" v-if="$slots.icon || icon">
      <slot name="icon">
        <var-icon class="var--flex" :name="icon" />
      </slot>
    </div>
    <div :class="n('content')">
      <div :class="classes(n('title'), [titleClass, titleClass])">
        <slot>{{ title }}</slot>
      </div>
      <div
        :class="classes(n('description'), [descriptionClass, descriptionClass])"
        v-if="$slots.description || description"
      >
        <slot name="description">
          {{ description }}
        </slot>
      </div>
    </div>
    <div :class="classes(n('extra'), [extraClass, extraClass])" v-if="$slots.extra">
      <slot name="extra" />
    </div>
  </div>
</template>

<script lang="ts">
import VarIcon from '../icon'
import { defineComponent } from 'vue'
import { props } from './props'
import { createNamespace } from '../utils/components'

const { n, classes } = createNamespace('cell')

export default defineComponent({
  name: 'VarCell',
  components: {
    VarIcon,
  },
  props,
  setup() {
    return {
      n,
      classes,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import './cell';
</style>
