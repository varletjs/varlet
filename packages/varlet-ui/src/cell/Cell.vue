<template>
  <div
    :class="classes(n(), [border, n('--border')], [onClick, n('--cursor')])"
    :style="borderOffsetStyles"
    v-ripple="{ disabled: !ripple }"
    @click="handleClick"
  >
    <slot name="icon">
      <div :class="classes(n('icon'), iconClass)" v-if="icon">
        <var-icon :name="icon" />
      </div>
    </slot>

    <div :class="n('content')">
      <slot>
        <div :class="classes(n('title'), titleClass)" v-if="title">
          {{ title }}
        </div>
      </slot>

      <slot name="description">
        <div :class="classes(n('description'), descriptionClass)" v-if="description">
          {{ description }}
        </div>
      </slot>
    </div>

    <div :class="classes(n('extra'), extraClass)" v-if="$slots.extra">
      <slot name="extra" />
    </div>
  </div>
</template>

<script lang="ts">
import VarIcon from '../icon'
import Ripple from '../ripple'
import { computed, defineComponent, type StyleValue, type ComputedRef } from 'vue'
import { props } from './props'
import { call, createNamespace } from '../utils/components'
import { toSizeUnit } from '../utils/elements'

const { n, classes } = createNamespace('cell')

export default defineComponent({
  name: 'VarCell',
  components: { VarIcon },
  directives: { Ripple },
  props,
  setup(props) {
    const borderOffsetStyles: ComputedRef<StyleValue> = computed(() => {
      if (props.borderOffset == null) {
        return {}
      }

      return {
        '--cell-border-left': toSizeUnit(props.borderOffset),
        '--cell-border-right': toSizeUnit(props.borderOffset),
      } as StyleValue
    })

    const handleClick = (e: Event) => {
      call(props.onClick, e)
    }

    return {
      n,
      classes,
      toSizeUnit,
      borderOffsetStyles,
      handleClick,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import './cell';
</style>
