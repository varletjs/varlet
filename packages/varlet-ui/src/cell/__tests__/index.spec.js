import Cell from '..'
import VarCell from '../Cell'
import VarIcon from '../../icon/Icon'
import { mount } from '@vue/test-utils'
import { createApp, h } from 'vue'

test('test cell plugin', () => {
  const app = createApp({}).use(Cell)
  expect(app.component(Cell.name)).toBeTruthy()
})

describe('test cell component props', () => {
  test('test cell title', () => {
    const wrapper = mount(VarCell, {
      props: {
        title: 'This is cell',
      },
    })

    expect(wrapper.find('.var-cell__title').text()).toBe('This is cell')
    wrapper.unmount()
  })

  test('test cell icon', () => {
    const wrapper = mount(VarCell, {
      props: {
        icon: 'fire',
      },
    })

    expect(wrapper.find('.var-icon').classes('var-icon-fire')).toBe(true)
    wrapper.unmount()
  })

  test('test cell desc', () => {
    const wrapper = mount(VarCell, {
      props: {
        desc: 'This is desc',
      },
    })

    expect(wrapper.find('.var-cell__desc').text()).toBe('This is desc')
    wrapper.unmount()
  })

  test('test cell border', async () => {
    const wrapper = mount(VarCell, {
      props: {
        border: true,
      },
    })

    expect(wrapper.classes('var-cell--border')).toBe(true)
    await wrapper.setProps({ border: false })
    expect(wrapper.classes('var-cell--border')).toBe(false)
    wrapper.unmount()
  })

  test('test cell icon class', () => {
    const wrapper = mount(VarCell, {
      props: {
        icon: 'fire',
        iconClass: 'test-icon-class',
      },
    })

    expect(wrapper.find('.var-cell__icon').classes()).toContain('test-icon-class')
    wrapper.unmount()
  })

  test('test cell title class', () => {
    const wrapper = mount(VarCell, {
      props: {
        title: 'This is Cell',
        titleClass: 'test-title-class',
      },
    })

    expect(wrapper.find('.var-cell__title').classes()).toContain('test-title-class')
    wrapper.unmount()
  })

  test('test cell desc class', () => {
    const wrapper = mount(VarCell, {
      props: {
        desc: 'This is desc',
        descClass: 'test-desc-class',
      },
    })

    expect(wrapper.find('.var-cell__desc').classes()).toContain('test-desc-class')
    wrapper.unmount()
  })

  test('test cell extra class', () => {
    const wrapper = mount(VarCell, {
      props: {
        extraClass: 'test-extra-class',
      },
      slots: {
        extra: 'information',
      },
    })

    expect(wrapper.find('.var-cell__extra').classes()).toContain('test-extra-class')
    wrapper.unmount()
  })
})

describe('test cell component slots', () => {
  test('test cell default slots', () => {
    const wrapper = mount(VarCell, {
      slots: {
        default: () => 'This is cell',
      },
    })

    expect(wrapper.find('.var-cell__title').element.textContent).toBe('This is cell')
    wrapper.unmount()
  })

  test('test cell icon slots', () => {
    const wrapper = mount(VarCell, {
      slots: {
        icon: h(VarIcon, { name: 'fire' }),
      },
    })

    expect(wrapper.find('.var-icon').classes('var-icon-fire')).toBe(true)
    wrapper.unmount()
  })

  test('test cell desc slots', () => {
    const wrapper = mount(VarCell, {
      slots: {
        desc: () => 'This is desc',
      },
    })

    expect(wrapper.find('.var-cell__desc').text()).toBe('This is desc')
    wrapper.unmount()
  })

  test('test cell extra slots', () => {
    const wrapper = mount(VarCell, {
      slots: {
        extra: 'information',
      },
    })

    expect(wrapper.find('.var-cell__extra').text()).toContain('information')
    wrapper.unmount()
  })
})
