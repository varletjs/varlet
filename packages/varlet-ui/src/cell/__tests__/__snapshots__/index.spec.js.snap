// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test cell example 1`] = `
"<div class=\\"app-type\\">基本使用</div>
<div class=\\"var-cell\\">
  <!--v-if-->
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
<div class=\\"var-cell\\">
  <!--v-if-->
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
<div class=\\"app-type\\">显示图标</div>
<div class=\\"var-cell\\">
  <div class=\\"var-cell__icon\\"><i class=\\"var-icon var-icon--set var-icon-fire var--flex\\" style=\\"transition: transform 0ms;\\"></i></div>
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell__extra\\"><i class=\\"var-icon var-icon--set var-icon-information\\" style=\\"transition: transform 0ms;\\"></i></div>
</div>
<div class=\\"var-cell\\">
  <div class=\\"var-cell__icon\\"><i class=\\"var-icon var-icon--set var-icon-fire var--flex\\" style=\\"transition: transform 0ms;\\"></i></div>
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <!--v-if-->
  </div>
  <div class=\\"var-cell__extra\\"><i class=\\"var-icon var-icon--set var-icon-information\\" style=\\"transition: transform 0ms;\\"></i></div>
</div>
<div class=\\"app-type\\">显示描述</div>
<div class=\\"var-cell\\">
  <div class=\\"var-cell__icon\\"><i class=\\"var-icon var-icon--set var-icon-fire var--flex\\" style=\\"transition: transform 0ms;\\"></i></div>
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <div class=\\"var-cell__desc\\">描述</div>
  </div>
  <div class=\\"var-cell__extra\\"><i class=\\"var-icon var-icon--set var-icon-information\\" style=\\"transition: transform 0ms;\\"></i></div>
</div>
<div class=\\"var-cell\\">
  <!--v-if-->
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <div class=\\"var-cell__desc\\">描述</div>
  </div>
  <!--v-if-->
</div>
<div class=\\"app-type\\">显示边框</div>
<div class=\\"var-cell var-cell--border\\">
  <!--v-if-->
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>
<div class=\\"var-cell var-cell--border\\">
  <!--v-if-->
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title\\">这是单元格</div>
    <!--v-if-->
  </div>
  <!--v-if-->
</div>"
`;

exports[`test cell props 1`] = `
"<div class=\\"var-cell var-cell--border\\">
  <div class=\\"var-cell__icon test-icon-class\\"><i class=\\"var-icon var-icon--set var-icon-fire var--flex\\" style=\\"transition: transform 0ms;\\"></i></div>
  <div class=\\"var-cell__content\\">
    <div class=\\"var-cell__title test-title-class\\">This is Cell</div>
    <div class=\\"var-cell__desc test-desc-class\\">This is desc</div>
  </div>
  <div class=\\"var-cell__extra test-extra-class\\">text</div>
</div>"
`;
